### PR TITLE
tools for locking / checking deprecated model elements

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/stormlib/model.py [hunk 1]
```diff
@@ -74,6 +75,92 @@
...
+                        if $layr.getPropCount($name, maxsize=1) {
+                            $lib.warn("Layer {iden} still contains {name}", iden=$layr.iden, name=$name)
+                            $ok = $lib.false
+                        }
+                    }
+                }
+
+                if (not $ok) {
+                    $lib.print("Whups! Your Cortex needs some tweaks to be fully future-model compliant.")
+                    $lib.print("See the docs at: https://foo")
```
Comment (mikemoritz) [inline synapse/lib/stormlib/model.py:156]: Just a reminder to update this :)

File: synapse/lib/stormlib/model.py [hunk 2]
```diff
@@ -74,6 +75,92 @@
...
+            ('name', {'help': 'The deprecated form or property name to lock.'}),
+            ('--unlock', {'help': 'Unlock rather than lock the deprecated property.', 'default': False, 'action': 'store_true'}),
+        ),
+        'storm': '''
+            init {
+                if $cmdopts.unlock {
+                    $lib.print("Unlocking: {name}", name=$cmdopts.name)
+                    $lib.model.deprecated.lock($cmdopts.name, $lib.false)
+                } else {
+                    if ($cmdopts.name = "*") {
```
Comment (mikemoritz) [inline synapse/lib/stormlib/model.py:91]: Might be worth adding this behavior in the help?

File: synapse/lib/stormlib/model.py [hunk 3]
```diff
@@ -356,3 +443,28 @@ async def _methEdgeList(self):
             retn.append((verb, kvdict))
 
         return retn
+
+@s_stormtypes.registry.registerLib
+class LibModelDeprecated(s_stormtypes.Lib):
```
Comment (mikemoritz) [inline synapse/lib/stormlib/model.py:448]: I assume the answer is no, but did you want any perms on modifying the locks?

File: synapse/lib/stormlib/model.py [hunk 4]
```diff
@@ -74,6 +75,92 @@
...
+                }
+
+                $lib.print("Checking for existance of deprecated model elements:")
+                for ($name, $locked) in $locks {
+
+                    $lib.print("{name}...", name=$name)
+
+                    for $layr in $lib.layer.list() {
+                        if $layr.getPropCount($name, maxsize=1) {
+                            $lib.warn("Layer {iden} still contains {name}", iden=$layr.iden, name=$name)
```
Comment (mikemoritz) [inline synapse/lib/stormlib/model.py:148]: Might be worth printing the count since you already have it?

Comment (invisig0th) [reply synapse/lib/stormlib/model.py:148]: Actually, i dont have the full count.  The maxsize=1 makes it so we just find 1 and return.

File: synapse/tests/test_lib_stormlib_model.py [hunk 5]
```diff
@@ -154,3 +154,57 @@ async def test_stormlib_model_edge(self):
...
+                await core.nodes('[ ou:org=* :sic=1234 ou:hasalias=($node.repr(), foobar) ]')
+
+                with self.raises(s_exc.NoSuchProp):
+                    await core.nodes('model.deprecated.lock newp:newp')
+
+                # lock a prop and a form/type
+                await core.nodes('model.deprecated.lock ou:org:sic')
+                await core.nodes('model.deprecated.lock ou:hasalias')
+
+                # TODO should these really be ReadOnly or something depr specific?
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_stormlib_model.py:174]: We should probably raise something more meaningful here

File: synapse/lib/stormtypes.py [hunk 6]
```diff
@@ -4447,11 +4455,27 @@ async def tostr(valu, noneok=False):
...
+        return await valu.iter()
+
+    return tuple(valu)
+
 async def tobool(valu, noneok=False):
 
     if noneok and valu is None:
         return None
 
+    if isinstance(valu, Prim):
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:4476]: we need to remove all the __bool__ implementations since they are now unused

File: synapse/cortex.py [hunk 7]
```diff
@@ -920,6 +920,22 @@ async def initServiceStorage(self):
...
+
+            form = self.model.form(name)
+            if form is not None:
+                form.locked = locked
+
+            prop = self.model.prop(name)
+            if prop is not None:
+                prop.locked = locked
+
+            _type = self.model.type(name)
```
Comment (vEpiphyte) [inline synapse/cortex.py:935]: its possible to have a form deprecated and the type itself not be deprecated. this name reuse would lock *both* of them.

File: synapse/datamodel.py [hunk 8]
```diff
@@ -253,7 +256,10 @@ def __init__(self, modl, name, info):
         self.props = {}     # name: Prop()
         self.refsout = None
 
-        if self.info.get('deprecated') or self.type.deprecated:
+        self.locked = False
+        self.deprecated = self.type.deprecated
```
Comment (vEpiphyte) [inline synapse/datamodel.py:260]: Is this matching form and type deprecation **together**? previously they were separate

File: synapse/lib/layer.py [hunk 9]
```diff
@@ -1422,7 +1422,7 @@ async def getTagCount(self, tagname, formname=None):
 
         return await self.layrslab.countByPref(abrv, db=self.bytag)
 
-    async def getPropCount(self, formname, propname=None):
+    async def getPropCount(self, formname, propname=None, maxsize=None):
```
Comment (vEpiphyte) [inline synapse/lib/layer.py:1425]: these changes are duplicated from #1931

Comment (vEpiphyte) [reply synapse/lib/layer.py:1425]: or are the changes there vestigial from this pr???

File: synapse/lib/stormlib/model.py [hunk 10]
```diff
@@ -356,3 +442,28 @@ async def _methEdgeList(self):
...
+    '''
+    _storm_lib_path = ('model', 'deprecated')
+
+    def getObjLocals(self):
+        return {
+            'lock': self._lock,
+            'locks': self._locks,
+        }
+
+    async def _locks(self):
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/model.py:459]: this is fine as a readonly function

File: synapse/lib/stormlib/model.py [hunk 11]
```diff
@@ -74,6 +75,91 @@
...
+                    for $layr in $lib.layer.list() {
+                        if $layr.getPropCount($name, maxsize=1) {
+                            $lib.warn("Layer {iden} still contains {name}", iden=$layr.iden, name=$name)
+                            $ok = $lib.false
+                        }
+                    }
+                }
+
+                if (not $ok) {
+                    $lib.print("Whups! Your Cortex needs some tweaks to be fully future-model compliant.")
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/model.py:155]: `Your Cortex currently has deprecated model elements in it.` or something. This message is unclear how to fix the problem or give hints towards how to fix it.

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**